### PR TITLE
Feature/snapsync server

### PR DESCRIPTION
**Pull request by asdacap**
- I'd like to merge snap sync server to master with its disabled so that I don't have to keep rebasing/merging.
- Think of it like a feature flag.
- Implementation is based on @tanishqjasoria's original implementation, but changed comparator to use TreePath and uses contextual visitor. It should work out of the box with any state db but might be slow.
- Here is how it perform. Metrics is nodes written measured from the syncing side. Only one peer and one thread. For reference, you'll need to sustain 1200k nodes per second to finish snap sync in 30 minute. Under most case, my SN850 will throttle it down to 900k nodes per second after about 5 minute.

| DB | Node per second |
| -- | -------|
| Geth | 600k-700k | 
| HalfPath | 170k-200k |
| Hash | 7k-8k |

- Hashdb work as expected as single thread IOPs is usually limited to 10k, compared to maximum nvme ssd throughput which is at least 500k. For the fun of it, when the serving node is using an optane drive, the node per second is 20k, which is disappointing. 
- It took about 27 hour to sync mainnet where the serving side uses optane drive. It'll probably take 2 to 3 days normally.
- This also means that turning it on by default is probably not a bad idea as it can't saturate any system resource except for background task queue.
- Note: Nethermind can't update snap sync pivot with only 1 node. Also, nethermind never use snap protocol in healing from another nethermind, so you'll need to modify something to test it.

## Changes

- Add snap server.
- Add tree path.

#### What types of changes does your code introduce?

- [X] New feature (a non-breaking change that adds functionality)

#### Requires testing

- [X] Yes
- [ ] No

#### If yes, did you write tests?

- [X] Yes
- [ ] No

#### Notes on testing

- Can confirm on halfpath can sync in about 3 hour.

### CODE REVIEW

File: src/Nethermind/Nethermind.Trie/PatriciaTree.cs [hunk 1]
```diff
@@ -1145,17 +1219,20 @@ public void Accept<TNodeContext>(ITreeVisitor<TNodeContext> visitor, Hash256 roo
                 // we introduced a notion of an account on the visit context level which should have no knowledge of account really
                 // but we know that we have multiple optimizations and assumptions on trees
                 ExpectAccounts = visitingOptions.ExpectAccounts,
-                MaxDegreeOfParallelism = visitingOptions.MaxDegreeOfParallelism
+                MaxDegreeOfParallelism = visitingOptions.MaxDegreeOfParallelism,
+                IsStorage = storageAddr != null
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Trie/PatriciaTree.cs:1223]: is not null (multiple places!)

File: src/Nethermind/Nethermind.Trie/PatriciaTree.cs [hunk 2]
```diff
@@ -1145,17 +1219,20 @@ public void Accept<TNodeContext>(ITreeVisitor<TNodeContext> visitor, Hash256 roo
...
+                if (storageRoot != null)
+                {
+                    rootHash = storageRoot;
+                }
+                else
                 {
-                    visitor.VisitMissingNode(default, rootHash, trieVisitContext);
-                    return;
+                    ReadOnlySpan<byte> bytes = Get(storageAddr.Bytes, rootHash);
+                    rootHash = AccountDecoder.Instance.DecodeStorageRootOnly(new RlpStream(bytes.ToArray()));
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Trie/PatriciaTree.cs:1235]: should not have to allocate array to decode something from it

File: src/Nethermind/Nethermind.Trie/PatriciaTree.cs [hunk 3]
```diff
@@ -1168,17 +1245,47 @@ public void Accept<TNodeContext>(ITreeVisitor<TNodeContext> visitor, Hash256 roo
...
+                if (rootHash != Keccak.EmptyTreeHash)
+                {
+                    rootRef = RootHash == rootHash ? RootRef : resolver.FindCachedOrUnknown(rootHash);
+                    if (!rootRef!.TryResolveNode(resolver))
+                    {
+                        visitor.VisitMissingNode(default, rootHash, trieVisitContext);
+                        return;
+                    }
+                }
+                rootRef?.Accept(visitor, default, resolver, trieVisitContext);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Trie/PatriciaTree.cs:1264]: duplicated logic

File: src/Nethermind/Nethermind.Trie/PatriciaTree.cs [hunk 4]
```diff
@@ -1168,17 +1245,47 @@ public void Accept<TNodeContext>(ITreeVisitor<TNodeContext> visitor, Hash256 roo
             ITrieNodeResolver resolver = TrieStore;
             if (flags != ReadFlags.None)
             {
-                resolver = new TrieNodeResolverWithReadFlags(TrieStore, flags);
+                resolver = new TrieNodeResolverWithReadFlags(resolver, flags);
             }
 
-            visitor.VisitTree(default, rootHash, trieVisitContext);
+            if (!visitor.IsFullDbScan)
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Trie/PatriciaTree.cs:1251]: this method from here is very hard to follow, multiple if's and returns, can you refactor for it to be more understandable?

File: src/Nethermind/Nethermind.Synchronization.Test/RangeQueryVisitorTests.cs [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+        };
+        inputStateTree.Accept(visitor, inputStateTree.RootHash, opt, storageAddr: account);
+        (var nodes, long _) = visitor.GetNodesAndSize();
+        nodes.Count.Should().Be(6);
+
+        int k = 0;
+        foreach (KeyValuePair<ValueHash256, byte[]> pair in nodes)
+        {
+            pair.Value.Should().BeEquivalentTo(TestItem.Tree.SlotsWithPaths[k + 0].SlotRlpValue);
+            k += 1;
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Synchronization.Test/RangeQueryVisitorTests.cs:118]: should be one assert

File: src/Nethermind/Nethermind.Synchronization.Test/RangeQueryVisitorTests.cs [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+    {
+        var startHash = new Hash256("0000000000000000000000000000000000000000000000000000000001113456");
+        var limitHash = new Hash256("0000000000000000000000000000000000000000000000000000000001123458");
+
+        using RangeQueryVisitor visitor = new(startHash, limitHash, false);
+        VisitingOptions opt = new()
+        {
+            ExpectAccounts = false,
+            MaxDegreeOfParallelism = 1,
+        };
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Synchronization.Test/RangeQueryVisitorTests.cs:56]: can we have one visiting options for all tests?

File: src/Nethermind/Nethermind.Synchronization.Test/SnapSync/SnapServerTest.cs [hunk 7]
```diff
@@ -0,0 +1,329 @@
...
+    {
+        MemDb stateDbServer = new();
+        MemDb codeDbServer = new();
+        TrieStore store = new(stateDbServer, LimboLogs.Instance);
+
+        StateTree tree = new(store, LimboLogs.Instance);
+
+        TestItem.Tree.FillStateTreeWithTestAccounts(tree);
+
+        SnapServer server = new(store.AsReadOnly(), codeDbServer, LimboLogs.Instance);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Synchronization.Test/SnapSync/SnapServerTest.cs:61]: A lot of reusable boilerplate setup code

File: src/Nethermind/Nethermind.Synchronization/SnapSync/SnapServer.cs [hunk 8]
```diff
@@ -0,0 +1,313 @@
...
+    }
+
+    public byte[][]? GetTrieNodes(PathGroup[] pathSet, in ValueHash256 rootHash, CancellationToken cancellationToken)
+    {
+        if (IsRootMissing(rootHash)) return Array.Empty<byte[]>();
+
+        _logger.Debug($"Trie node {pathSet.Length}");
+        // TODO: use cache to reduce node retrieval from disk
+        int pathLength = pathSet.Length;
+        List<byte[]> response = new(pathLength);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Synchronization/SnapSync/SnapServer.cs:98]: could we use ArrayPoolList, that would get disposed after the message is sent?

File: src/Nethermind/Nethermind.Synchronization/SnapSync/SnapServer.cs [hunk 9]
```diff
@@ -0,0 +1,313 @@
...
+                    }
+                    catch (MissingTrieNodeException)
+                    {
+                        TrackTrieNodeException(rootHash);
+                        abort = true;
+                    }
+                    break;
+            }
+        }
+        return response.ToArray();
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Synchronization/SnapSync/SnapServer.cs:150]: regardless we have List and the ToArray, ArrayPoolList would at least reduce allocations a bit

File: src/Nethermind/Nethermind.Synchronization/SnapSync/SnapServer.cs [hunk 10]
```diff
@@ -0,0 +1,313 @@
...
+    }
+
+    public byte[][] GetByteCodes(IReadOnlyList<ValueHash256> requestedHashes, long byteLimit, CancellationToken cancellationToken)
+    {
+        long currentByteCount = 0;
+        List<byte[]> response = new(requestedHashes.Count);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Synchronization/SnapSync/SnapServer.cs:156]: same here

File: src/Nethermind/Nethermind.Synchronization/SnapSync/SnapServer.cs [hunk 11]
```diff
@@ -0,0 +1,313 @@
...
+
+            if (cancellationToken.IsCancellationRequested)
+            {
+                break;
+            }
+
+            // TODO: handle empty code in the DB itself?
+            if (codeHash.Bytes.SequenceEqual(Keccak.OfAnEmptyString.Bytes))
+            {
+                response.Add(Array.Empty<byte>());
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Synchronization/SnapSync/SnapServer.cs:180]: missing break?

File: src/Nethermind/Nethermind.Synchronization/SnapSync/SnapServer.cs [hunk 12]
```diff
@@ -0,0 +1,313 @@
...
+        StateTree tree = new(_storeWithReadFlag, _logManager);
+
+        ValueHash256 startingHash1 = startingHash ?? ValueKeccak.Zero;
+        ValueHash256 limitHash1 = limitHash ?? ValueKeccak.MaxValue;
+        if (limitHash1 == ValueKeccak.Zero)
+        {
+            limitHash1 = ValueKeccak.MaxValue;
+        }
+
+        List<PathWithStorageSlot[]> responseNodes = new(accounts.Length);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Synchronization/SnapSync/SnapServer.cs:225]: same optimization potential here

File: src/Nethermind/Nethermind.Trie.Test/NibbleTests.cs [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+
+public class NibbleTests
+{
+    private readonly byte[][] _hexEncoding = {
+        new byte[]{},
+        new byte[]{16},
+        new byte[]{1, 2, 3, 4, 5},
+        new byte[]{0, 1, 2, 3, 4, 5},
+        new byte[]{15, 1, 12, 11, 8, 16},
+        new byte[]{0, 15, 1, 12, 11, 8, 16},
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Trie.Test/NibbleTests.cs:17]: new collection initializers?

File: src/Nethermind/Nethermind.Trie.Test/TreePathTests.cs [hunk 14]
```diff
@@ -0,0 +1,168 @@
...
+    }
+
+    [Test]
+    public void TestIndexWrite()
+    {
+        TreePath path = CreateFullTreePath();
+
+        string asHex = path.Span.ToHexString();
+        asHex.Should().Be("0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef");
+    }
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Trie.Test/TreePathTests.cs:31]: same test?

File: src/Nethermind/Nethermind.Synchronization/SnapSync/SnapServer.cs (hunk 8 above)
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Synchronization/SnapSync/SnapServer.cs]: used ArrayPool for local construction, but could we keep it until we send message and then dispose?

File: src/Nethermind/Nethermind.Network/IProtocolsManager.cs [hunk 15]
```diff
@@ -14,7 +14,8 @@ public interface IProtocolsManager
         void AddSupportedCapability(Capability capability);
         void RemoveSupportedCapability(Capability capability);
         void SendNewCapability(Capability capability);
-        void AddProtocol(string code, Func<ISession, IProtocolHandler> factory);
+        void AddProtocol(string code, Func<ISession, int, IProtocolHandler> factory);
+        void InitSatelliteProtocol(ISession session, ProtocolHandlerBase handler);
```
[the remark this answers is not in the document]
Comment (asdacap) [reply src/Nethermind/Nethermind.Network/IProtocolsManager.cs:18]: I'll rework this from outside.

File: src/Nethermind/Nethermind.Blockchain/Synchronization/ISyncConfig.cs [hunk 16]
```diff
@@ -139,4 +139,7 @@ public interface ISyncConfig : IConfig
 
     [ConfigItem(Description = "Interval, in seconds, between `malloc_trim` calls during sync.", DefaultValue = "300", HiddenFromDocs = true)]
     public int MallocTrimIntervalSec { get; set; }
+
+    [ConfigItem(Description = "_Experimental._ Whether to enable snap serving. WARNING: Very slow on hash db layout.", DefaultValue = "false", HiddenFromDocs = true)]
+    bool SnapServe { get; set; }
```
[the remark this answers is not in the document]
Comment (asdacap) [reply src/Nethermind/Nethermind.Blockchain/Synchronization/ISyncConfig.cs:144]: Dunno, which one do you prefer?

File: src/Nethermind/Nethermind.Network/P2P/Subprotocols/Snap/Messages/ByteCodesMessageSerializer.cs [hunk 17]
```diff
@@ -43,7 +43,7 @@ public static (int contentLength, int codesLength) GetLength(ByteCodesMessage me
                 codesLength += Rlp.LengthOf(message.Codes[i]);
             }
 
-            return (codesLength + Rlp.LengthOf(message.RequestId), codesLength);
+            return (Rlp.LengthOfSequence(codesLength) + Rlp.LengthOf(message.RequestId), codesLength);
```
[the remark this answers is not in the document]
Comment (asdacap) [reply src/Nethermind/Nethermind.Network/P2P/Subprotocols/Snap/Messages/ByteCodesMessageSerializer.cs:46]: Dunno. I do recall trying to revert some of the serializer change and got invalid hash. Not sure if this is the one. @tanishqjasoria  do you remember?

File: src/Nethermind/Nethermind.Network/P2P/Subprotocols/Snap/SnapProtocolHandler.cs [hunk 18]
```diff
@@ -274,5 +364,26 @@ private static PathGroup[] GetPathGroups(AccountsToRefreshRequest request)
                 static (request) => request.ToString(),
                 token);
         }
+
+        protected void ScheduleSyncServe<TReq, TRes>(TReq request, Func<TReq, CancellationToken, Task<TRes>> fulfillFunc) where TRes : P2PMessage
```
[the remark this answers is not in the document]
Comment (asdacap) [reply src/Nethermind/Nethermind.Network/P2P/Subprotocols/Snap/SnapProtocolHandler.cs]: Other response is handled in `SyncPeerProtocolBase`. Snap sync does not implement it, so it need to copy the method.

File: src/Nethermind/Nethermind.Trie/NibbleExtensions.cs [hunk 19]
```diff
@@ -142,6 +143,34 @@ public static byte[] ToBytes(ReadOnlySpan<byte> nibbles)
             return bytes;
         }
 
+        public static byte[] CompactToHexEncode(byte[] compactPath)
```
[the remark this answers is not in the document]
Comment (asdacap) [reply src/Nethermind/Nethermind.Trie/NibbleExtensions.cs:146]: True

[the remark this answers is not in the document]
Comment (asdacap) [reply src/Nethermind/Nethermind.Trie/NibbleExtensions.cs:146]: Ah, it always chop it at the front.  Probably can stackalloc intead of pool though. @benaadams do stackalloc  have limit?

File: src/Nethermind/Nethermind.Trie/PatriciaTree.cs [hunk 20]
```diff
@@ -379,6 +379,49 @@ public virtual ReadOnlySpan<byte> Get(ReadOnlySpan<byte> rawKey, Hash256? rootHa
...
+        [DebuggerStepThrough]
+        public byte[]? GetNodeByKey(Span<byte> rawKey, Hash256? rootHash = null)
+        {
+            byte[] array = null;
+            try
+            {
+                int nibblesCount = 2 * rawKey.Length;
+                Span<byte> nibbles = rawKey.Length <= 64
+                    ? stackalloc byte[nibblesCount]
+                    : array = ArrayPool<byte>.Shared.Rent(nibblesCount);
```
[the remark this answers is not in the document]
Comment (asdacap) [reply src/Nethermind/Nethermind.Trie/PatriciaTree.cs:408]: Probably missed as the original code is probably before the `MaxKeyStackAlloc` exist.

File: src/Nethermind/Nethermind.Trie/PatriciaTree.cs [hunk 21]
```diff
@@ -840,7 +893,7 @@ private void ConnectNodes(TrieNode? node, in TraverseContext traverseContext)
 
             if (childNode is null)
             {
-                if (traverseContext.IsRead)
+                if (traverseContext.IsRead || traverseContext.IsNodeRead)
```
[the remark this answers is not in the document]
Comment (asdacap) [reply src/Nethermind/Nethermind.Trie/PatriciaTree.cs:896]: Its more like `CanReturn` instead. Need a better term for the combination of these two. `IsRead` -> `IsValueRead` and the combination is `IsRead`?

File: src/Nethermind/Nethermind.Trie/NibbleExtensions.cs (hunk 19 above)
[the remark this answers is not in the document]
Comment (benaadams) [reply src/Nethermind/Nethermind.Trie/NibbleExtensions.cs:146]: Depends on how deep the stack goes and calls on calls (and whether they are also stackallocing)' should be able to do say 256 ok though; and then flip to arraypool rent as stack overflow is very nasty

File: src/Nethermind/Nethermind.Trie/PatriciaTree.cs (hunk 20 above)
[the remark this answers is not in the document]
Comment (benaadams) [reply src/Nethermind/Nethermind.Trie/PatriciaTree.cs:408]: Fixed size stackalloc is better otherwise it has to do stack probes based on unknown length

File: src/Nethermind/Nethermind.Synchronization/SnapSync/SnapServer.cs (hunk 11 above)
Comment (asdacap) [reply src/Nethermind/Nethermind.Synchronization/SnapSync/SnapServer.cs:173]: No, looks like we can handle skipped entries. Geth usually does not include null in their list, and just skip them, so I assume they work the same.

File: src/Nethermind/Nethermind.Trie.Test/TreePathTests.cs (hunk 14 above)
Comment (asdacap) [reply src/Nethermind/Nethermind.Trie.Test/TreePathTests.cs:31]: Probably a mistake when consolidating common code.

**Comment by asdacap** [comment]
I prefer early return like this.

```
if (a) break;
if (b) break;
if (c) break;
// logic
```

Easier to read as the code is kept to the left. 
Its easy to know how many condition does it need to pass. 
Its easy to know all of these condition are of the same abstraction level, as they are on the same vertical line.
If a condition is removed or added, the diff is easy to review as it add line or remove line without messing with other condition. You can also add a more complicated, multiline condition easier without messing with the logic code.

**Comment by LukaszRozmej** [comment]
Maybe out of scope of this PR, but can we make a refactor to use Pooled collections in p2p messages (inbound and outbound) and then dispose the massages after read/sent to release the pooled collections back?

File: src/Nethermind/Nethermind.Network/IProtocolsManager.cs [hunk 22]
```diff
@@ -14,7 +14,8 @@ public interface IProtocolsManager
         void AddSupportedCapability(Capability capability);
         void RemoveSupportedCapability(Capability capability);
         void SendNewCapability(Capability capability);
-        void AddProtocol(string code, Func<ISession, IProtocolHandler> factory);
+        void AddProtocol(string code, Func<ISession, int, IProtocolHandler> factory);
```
[the remark this answers is not in the document]
Comment (asdacap) [reply src/Nethermind/Nethermind.Network/IProtocolsManager.cs]: Moved network refactor to #6730. I'll just pass in snap server to protocol manager.

**Comment by asdacap** [comment]
Yea, I thought of implementing `IDispose` to `P2PMessage`

**Review by LukaszRozmej — APPROVED** [review]
Maybe let's wait for more reviews too

File: src/Nethermind/Nethermind.Blockchain/Synchronization/ISyncConfig.cs (hunk 16 above)
[the remark this answers is not in the document]
Comment (kamilchodola) [reply src/Nethermind/Nethermind.Blockchain/Synchronization/ISyncConfig.cs:144]: SnapServingEnabled? :D

[the remark this answers is not in the document]
Comment (benaadams) [reply src/Nethermind/Nethermind.Blockchain/Synchronization/ISyncConfig.cs:144]: `SnapServerNode` would go with `NonValidatorNode`

File: src/Nethermind/Nethermind.Synchronization/Peers/SyncPeerExtensions.cs [hunk 23]
```diff
@@ -29,13 +29,6 @@ public static bool SupportsAllocation(this PeerInfo peerInfo, AllocationContexts
                 }
             }
 
-            // check if peer supports snap sync
-            if ((contexts & AllocationContexts.Snap) != 0)
-            {
-                // TODO: Remove when Nethermind implements snap server
```
Comment (benaadams) [inline src/Nethermind/Nethermind.Synchronization/Peers/SyncPeerExtensions.cs:35]: 🚀

File: src/Nethermind/Nethermind.Trie/NibbleExtensions.cs [hunk 24]
```diff
@@ -142,6 +145,40 @@ public static byte[] ToBytes(ReadOnlySpan<byte> nibbles)
...
+            {
+                return compactPath;
+            }
+            int nibblesCount = compactPath.Length * 2 + 1;
+            byte[]? array = null;
+            Span<byte> nibbles = nibblesCount < StackAllocLengthLimit
+                ? stackalloc byte[nibblesCount]
+                : array ??= ArrayPool<byte>.Shared.Rent(nibblesCount);
+
+            try
```
Comment (benaadams) [inline src/Nethermind/Nethermind.Trie/NibbleExtensions.cs:160]: Don't need try/finally here, if throw an exception doesn't matter too much if don't return array to the shared array pool

File: src/Nethermind/Nethermind.Trie/PatriciaTree.cs [hunk 25]
```diff
@@ -379,6 +379,50 @@ public virtual ReadOnlySpan<byte> Get(ReadOnlySpan<byte> rawKey, Hash256? rootHa
...
+                throw new TrieException(
+                    $"Failed to load by path {nibbles.ToHexString()} from root hash {rootHash ?? RootHash}.", e);
+            }
+        }
+
+        [DebuggerStepThrough]
+        public byte[]? GetNodeByKey(Span<byte> rawKey, Hash256? rootHash = null)
+        {
+            byte[] array = null;
+            try
```
Comment (benaadams) [inline src/Nethermind/Nethermind.Trie/PatriciaTree.cs:403]: Same here, no need for try finally

File: src/Nethermind/Nethermind.Trie/PatriciaTree.cs [hunk 26]
```diff
@@ -379,6 +379,50 @@ public virtual ReadOnlySpan<byte> Get(ReadOnlySpan<byte> rawKey, Hash256? rootHa
...
+            {
+                int nibblesCount = nibbles.Length;
+                CappedArray<byte> result = Run(nibbles, nibblesCount, Array.Empty<byte>(), false, startRootHash: rootHash,
+                    isNodeRead: true);
+                return result.ToArray() ?? Array.Empty<byte>();
+            }
+            catch (TrieException e)
+            {
+                throw new TrieException(
```
Comment (benaadams) [inline src/Nethermind/Nethermind.Trie/PatriciaTree.cs:394]: Move this out to local function; new exception + string interpolation = very bloated stack clearing on method entry even if exception not thrown

File: src/Nethermind/Nethermind.Trie/PatriciaTree.cs [hunk 27]
```diff
@@ -379,6 +379,50 @@ public virtual ReadOnlySpan<byte> Get(ReadOnlySpan<byte> rawKey, Hash256? rootHa
...
+                        : array = ArrayPool<byte>.Shared.Rent(nibblesCount))
+                    [..nibblesCount]; // Slice to exact size;
+                Nibbles.BytesToNibbleBytes(rawKey, nibbles);
+                CappedArray<byte> result = Run(nibbles, nibblesCount, Array.Empty<byte>(), false, startRootHash: rootHash,
+                    isNodeRead: true);
+                return result.ToArray() ?? Array.Empty<byte>();
+            }
+            catch (TrieException e)
+            {
+                throw new TrieException(
```
Comment (benaadams) [inline src/Nethermind/Nethermind.Trie/PatriciaTree.cs:417]: Move this out to local function for reason above

File: src/Nethermind/Nethermind.Trie/VisitingOptions.cs [hunk 28]
```diff
@@ -39,13 +39,12 @@ public int MaxDegreeOfParallelism
...
-            if (rawMaxDegreeOfParallelism == 0)
-                return Math.Max(Environment.ProcessorCount / 4, 1);
-            if (rawMaxDegreeOfParallelism <= -1)
-                return Environment.ProcessorCount;
-            return rawMaxDegreeOfParallelism;
-        }
+        public static int AdjustMaxDegreeOfParallelism(int rawMaxDegreeOfParallelism) =>
+            rawMaxDegreeOfParallelism switch
+            {
+                0 => Math.Max(Environment.ProcessorCount / 4, 1),
```
Comment (benaadams) [inline src/Nethermind/Nethermind.Trie/VisitingOptions.cs:45]: If you surface `Nethermind\Nethermind.Init\Cpu\RuntimeInformation.cs` in some way (perhaps via config) then can use `RuntimeInformation.GetCpuInfo()?.PhysicalCoreCount /2 ?? Environment.ProcessorCount / 4` if that's what you were going for here?

**Review by benaadams — APPROVED** [review]

File: src/Nethermind/Nethermind.Network/P2P/Subprotocols/Snap/Messages/ByteCodesMessageSerializer.cs (hunk 17 above)
[the remark this answers is not in the document]
Comment (tanishqjasoria) [reply src/Nethermind/Nethermind.Network/P2P/Subprotocols/Snap/Messages/ByteCodesMessageSerializer.cs:46]: i remember that there was a issue with one of the serializer while implmenting the server.

[the remark this answers is not in the document]
Comment (tanishqjasoria) [reply src/Nethermind/Nethermind.Network/P2P/Subprotocols/Snap/Messages/ByteCodesMessageSerializer.cs:46]: yes  - https://github.com/NethermindEth/nethermind/pull/4742/files#diff-51423375bb5d0c69fcddb8758c53c0cfb549fdc0e45bf0e57bb671b723bda559R46

File: src/Nethermind/Nethermind.Blockchain/Synchronization/ISyncConfig.cs (hunk 16 above)
[the remark this answers is not in the document]
Comment (asdacap) [reply src/Nethermind/Nethermind.Blockchain/Synchronization/ISyncConfig.cs:144]: I'll set to `SnapServingEnabled`.

**Review by benaadams — APPROVED** [review]

**Comment by LukaszRozmej** [comment]
Btw do they pass hive snap tests? https://hivetests.ethdevops.io/suite.html?suiteid=1707924603-d4f31d61df08ea6e9724ae917ad13ac3.json&suitename=snap

**Comment by asdacap** [comment]
Screenshot_2024-02-19_08-23-23

Got some kind of forkid failure.
